### PR TITLE
Feature/retryablehttp

### PR DESCRIPTION
**Pull request by GWeggemans**

### CODE REVIEW

File: spec/components/retryer/RetryerSpec.ts [hunk 1]
```diff
@@ -41,15 +41,15 @@ describe('A Retryer', () => {
...
-    const retryer = new Retryer(statsDMock, retryableMock, 5, 1, 1, 1);
+    const retryer = new Retryer(statsDMock, retryableMock, 5, 1, 1, 1, ['service:intershop']);
 
     await retryer.execute();
 
     expect(retryableMock.attempt).toHaveBeenCalledTimes(3);
     expect(tries).toEqual(3);
     expect(statsDMock.increment).toHaveBeenCalledTimes(2);
-    expect((statsDMock.increment as jasmine.Spy).calls.argsFor(0)).toEqual(['retryable.retries', jasmine.any(Number), ['tags:one', 'retryAttempt:1']]);
-    expect((statsDMock.increment as jasmine.Spy).calls.argsFor(1)).toEqual(['retryable.retries', jasmine.any(Number), ['tags:one', 'retryAttempt:2']]);
```
Comment (yisonglin) [inline spec/components/retryer/RetryerSpec.ts]: what does this do?

File: spec/components/retryer/RetryerSpec.ts [hunk 2]
```diff
@@ -30,7 +30,7 @@ describe('A Retryer', () => {
 
     retryableMock.attempt = () => Promise.resolve(true);
 
-    const retryer = new Retryer(statsDMock, retryableMock, 2, 1, 1, 1);
+    const retryer = new Retryer(statsDMock, retryableMock, 2, 1, 1, 1, ['service:intershop']);
```
Comment (ryreitsma) [inline spec/components/retryer/RetryerSpec.ts]: It's not very readable to have arguments like this in a constructor. As a consumer of this code, I can't see what it does, what does `2,1,1,1` do?

Maybe have a look at using named parameters in this constructor? I think Typescript has a way to do this, would result in something like:
```javascript
const retryer = new Retryer({statsD: statsDMock, retryable: retryableMock, maxAttempts: 2, minInterval: 1, maxInterval: 1, jitter: 1, defaultTags: ['service:intershop']});
```
I've seen that you can also set defaults in the constructor so you won't have to specify all of the arguments. I think most users of this constructor would be fine with having the intervals and jitter set to 1 by default

Comment (GWeggemans) [reply spec/components/retryer/RetryerSpec.ts]: Thanks, I agree

Typescript does not support "real" named parameters in the constructor:
https://github.com/Microsoft/TypeScript/issues/467

When using an object like the example above, you can't use default constructor values anymore. I will play around looking nice solution.

Comment (GWeggemans) [reply spec/components/retryer/RetryerSpec.ts]: image
Also, maybe webstorm is making me lazy :)

Comment (GWeggemans) [reply spec/components/retryer/RetryerSpec.ts]: Fixed

File: src/components/httprequest/RetryableHTTPRequest.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  public constructor(request: HTTPRequest) {
+    super(request);
+  }
+
+  public get tags(): string[] {
+    const tags: Record<string, unknown> = JSON.parse(JSON.stringify(this.options));
+
+    delete tags.body;
+
+    if (this.requestResult) {
```
Comment (FBajo) [inline src/components/httprequest/RetryableHTTPRequest.ts]: You could refactor the `tags` method into less responsibilities. Perhaps adding a (private) `validateRequestResult` method underneath

Comment (FBajo) [reply src/components/httprequest/RetryableHTTPRequest.ts]: Should still pass the tests of course

File: src/components/httprequest/RetryableHTTPRequest.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        tags.statusCode = `${this.requestResult.statusCode}`;
+      }
+    } else {
+      tags.result = 'noattempt';
+    }
+
+    return objectToTags(tags);
+  }
+
+  public async attempt(): Promise<boolean> {
```
Comment (FBajo) [inline src/components/httprequest/RetryableHTTPRequest.ts]: This method name could be a bit more specific. Attempt what? And also it seems to be returning a boolean, so I would continue with your `is` prefix style

Comment (FBajo) [reply src/components/httprequest/RetryableHTTPRequest.ts]: Hmm ok I see, `attempt` can be seen as a verb. Maybe `attemptToRetry` ? I'll leave it up to you on this one

File: src/components/httprequest/RetryableHTTPRequest.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    return this.requestResult;
+  }
+
+  public getLogInformation(): object {
+    return this.options;
+  }
+
+  private isRetryable(requestResult: HTTPRequestResponse | Error): boolean {
+    if (requestResult instanceof Error) {
+      return requestResult.message !== 'Error: ESOCKETTIMEDOUT' &&
```
[the remark this answers is not in the document]
Comment (GWeggemans) [reply src/components/httprequest/RetryableHTTPRequest.ts]: Improved it

**Pull request merged**